### ISSUE TITLE
Feature request: support mixed precision for GradCam(PlusPlus)

### ISSUE THREAD

**Issue by bersbersbers** [keisen/tf-keras-vis#41]
This short example shows that `GradCam(PlusPlus)` fails when the network is trained with mixed precision:

```Python
import numpy as np
import tensorflow.keras as keras
from tensorflow.keras.mixed_precision import experimental as mixed_precision
from tf_keras_vis.gradcam import Gradcam, GradcamPlusPlus
from tf_keras_vis.saliency import Saliency

policy = mixed_precision.Policy("mixed_float16")
mixed_precision.set_policy(policy)

model = keras.applications.MobileNet()
input_data = np.empty(model.input.shape[1:])


def model_modifier(model):
    model.layers[-1].activation = keras.activations.linear
    return model


def loss(output):
    return [o[0] for o in output]


saliency = Saliency(model, model_modifier, clone=False)

# works
saliency(loss, input_data)

# works
saliency(loss, input_data, smooth_samples=20)

# does not work: "array type dtype('float16') not supported"
gradcam = Gradcam(model, model_modifier, clone=False)
gradcam(loss, input_data, penultimate_layer=-1)

# does not work: "array type dtype('float16') not supported"
gradcampp = GradcamPlusPlus(model, model_modifier, clone=False)
gradcampp(loss, input_data, penultimate_layer=-1)
```

It works fine when you comment out `mixed_precision.set_policy(policy)`.

**Comment by keisen** [comment]
Hi, @bersbersbers . Thank you for your reporting!
I'm going to fix this issue immediately.

Thanks!

**Comment by bersbersbers** [comment]
Thanks, that was fast! Now that this is fixed, I am getting further errors unfortunately:

> Exception has occurred: InvalidArgumentError
> cannot compute Mul as input #1(zero-based) was expected to be a float tensor but is a half tensor [Op:Mul]

I'll see if I can reduce this in an example, too.

**Comment by bersbersbers** [comment]
Here's a minimal example that is giving the above error:

```
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import applications, layers
from tensorflow.keras.mixed_precision import experimental as mixed_precision
from tf_keras_vis.gradcam import GradcamPlusPlus

policy = mixed_precision.Policy("mixed_float16")
mixed_precision.set_policy(policy)

base_model = applications.MobileNet(
    input_shape=[224, 244, 3],
    include_top=False,
    weights=None,
)
layer = base_model.output
layer = layers.Flatten(name="flatten")(layer)
layer = layers.Dense(2, dtype=tf.float32)(layer)
model = keras.models.Model(inputs=base_model.input, outputs=layer)
data = np.empty(model.input.shape[1:])


def model_modifier(model):
    model.layers[-1].activation = keras.activations.linear
    return model


def loss(output):
    return [o[0] for o in output]


GradcamPlusPlus(model, model_modifier)(loss, data)
print("Done.")
```

I reduced it quite a bit, so maybe this network does not make sense any more from a learning perspective. But it technically works without `"mixed_float16"`, and fails with `"mixed_float16"`.

The problem is the data type of the output layer (`dtype=tf.float32`). I don't remember why, but I added this in the same commit in which I enabled mixed precision, so it must be due to some incompatibility elsewhere (maybe saving/loading the model). It's on my to-do-list to remove this, but this requires some testing and if `tf-keras-vis` could be made to support this use case, too, I would be very happy :)

**Comment by keisen** [comment]
Thank you for your additional reporting.

Indeed, I believe that the case which the data type of model-outputs and gradients are different is rare. However, ...

> It's maybe safe to cast the scalar score to the dtype of grads?

On the other hands, I also believe that this idea is reasonable to deal with this problem safely.
So we decided we implement it in next release.

Thanks!

**Comment by bersbersbers** [comment]
Looking forward to that, thanks again!

**Comment by bersbersbers** [comment]
I found some additional information on why I added `dtype=tf.float32` in my example here:
> A softmax activation at the end of the model should be float32.

https://www.tensorflow.org/guide/mixed_precision#building_the_model

**Comment by bersbersbers** [comment]
This is not fixed in 66132db3, even without the use of `experimental`:

```python
# pip install tensorflow==2.4.1 git+https://github.com/keisen/tf-keras-vis@66132db3
import tensorflow as tf
from tf_keras_vis import gradcam

tf.keras.mixed_precision.set_global_policy("mixed_float16")
base_model = tf.keras.applications.MobileNet(
    include_top=False,
    input_shape=(32, 32, 3),
    weights=None,
)
layer = base_model.output
layer = tf.keras.layers.Flatten(name="flatten")(layer)
layer = tf.keras.layers.Dense(2, dtype=tf.float32)(layer)
model = tf.keras.models.Model(inputs=base_model.input, outputs=layer)
data = tf.zeros(model.input.shape[1:])
loss = lambda output: sum(output)
gradcam.Gradcam(model)(loss, data)
print("Done.")
```

> `RuntimeError: array type dtype('float16') not supported`

**Comment by keisen** [comment]
Hi @bersbersbers . Thank you for pointing it out.
I've fixed it (See for details: 0dd534ba61e4e91a208a4650aa5a05c65022e62f and cf4ba3aa9d0d5e9846e28ce4f745235772549381).

Thanks!
